docs: docs for remote joins (close #4911) (#5132)
hasura/graphql-engine: docs/graphql/manual/schema/relationships/architecture.rst

> .. meta::

Reviewer: Can we rename this file to `remote-joins-architecture` ?

   :description: Security, performance, schema integrity, and data federation with Hasura remote joins
   :keywords: hasura, docs, data federation, remote relationship, remote join, remote schema

.. _remote_joins_architecture:

> Architecture of remote joins

Reviewer 2: We need to think about the title since it's not only for remote joins and it's also not really about architecture. @rikinsk and me were thinking about it and we're like 🤯 not able to come up with anything better yet :D

Author: I based this on @tirumaraiselvan 's talk on remote joins. I believe the concept of the join applies at the data federation level, e.g. you can also join with a field returned from an action.

> Architecture of remote joins

Reviewer 2: Also, we need to be consistent with the naming: either "Remote joins" or "Remote relationships". @rikinsk and me think it should be "Remote relationships" since then the wording is consistent with the rest of the headings 🙂

> Architecture of remote relationships

Reviewer: Let's call it `Architecture of Remote Joins`

Author: Lol that's what I had originally, cc @marionschleifer @rikinsk

============================

.. contents:: Table of contents
  :backlinks: none
  :depth: 1
  :local:

> 

Reviewer 2: Can we add a section called `Introduction` explaining what this page is about?

Introduction
------------

> Hasura has a unique solution to the data federation problem, in that join, authorization, and consistency checks all happen at the Hasura layer, allowing underlying data sources and APIs to evolve independently. With this model, your applications have a unified API with which to harness the full power of your data.

Reviewer: ```suggestion
Hasura has a unique way of bringing data sources together and solving the data federation problem. This is done via Remote Joins wherein you can join data between data sources, similar to table relationships.  

With Remote joins, the join, authorization, and consistency checks (of added sources) all happen at the Hasura layer via metadata. This allows underlying data sources and APIs to evolve independently. Your applications have a unified API to query the full data landscape in your org.  
```

Security
--------

> Hasura's :ref:`authorization model <authorization>` gives you role-based access control. This extends to actions and remote schemas. In actions, you can go to the "Permissions" tab and choose which fields to expose for which roles. In remote schemas, Hasura will :ref:`forward <schema_auth>` your session variables to your custom authorization implementation. More fine-grained role-based exposure of remote schema fields is in the works.

Reviewer: ```suggestion
Hasura's :ref:`authorization model <authorization>` gives you role-based access control. This extends to actions and remote schemas as well. In actions, you can go to the "Permissions" tab and choose which fields to expose for which roles. In remote schemas, Hasura will :ref:`forward <schema_auth>` your session variables which cane be used to implement authorization in your remote schema. Role-based permissions on remote schemas is in the works.
```

Performance
-----------

> Hasura strives for optimal performance. It creates an execution plan, which pushes down most of the heavy-lifting involved in a query, sending a single efficient query to your database or your remote GraphQL endpoints, avoiding the `n+1 problem <https://hasura.io/learn/graphql/intro-graphql/graphql-server/>`__ in the process.

Reviewer: ```suggestion
Hasura strives for optimal performance. It creates an efficient execution plan which pushes down most of the heavy-lifting involved to underlying sources. For example, Hasura creates a single efficient query to your database and batches calls to Remote Schemas to avoid the `n+1 problem <https://hasura.io/learn/graphql/intro-graphql/graphql-server/>`__ . More improvements to performance are upcoming.
```

Schema integrity & consistency
------------------------------

> Hasura has metadata consistency checks at every level. Whenever you add a table, remote schema, or action, Hasura makes sure that the graph that is exposed is consistent, and that all the relationships make sense at a type level. This helps with setting up tooling, such as integration tests in your CI for safe changes.

Reviewer: ```suggestion
Hasura has metadata consistency checks at every level. Whenever you add a table, remote schema, or action, Hasura makes sure that the graph that is exposed is consistent, and that all the relationships make sense at a type level. This helps you in creating robust workflows to test changes in your CI and making safe deployments.
```
